### ISSUE TITLE
Activity Image not working

### ISSUE THREAD

**Issue by palittle** [HTBox/allReady#550]
On the Create Activity page `http://allready-d.azurewebsites.net/Admin/Activity/Create/7` I clicked the Choose File button and opened an image and it seemed to upload but when I clicked the Create (Activity) button the image file was no longer there.

**Comment by palittle** [comment]
Thought this wasn't working at all but what happened was that the date was invalid and therefore when I clicked Create it didn't save and the image was for some reason taken away.

**Comment by palittle** [comment]
Not showing now... unless someone removed it, it should be there.

**Comment by dangle1** [comment]
I'd like to take a look at this.

**Comment by dangle1** [comment]
@palittle,

just to be clear: the expectation here is that even if there is a validation error on the page (an invalid date in your case), and the browser comes back to the same page, the image should still be selected and the user should not have to redo the image upload process, correct?

**Comment by palittle** [comment]
Yes - to not have to choose the image again after a validation error would definitely help. The image was visible after I created the activity but then later when I left the page and returned it seemed to have disappeared. Not sure if this still happens as I can't seem to get into the activity page from the campaign page any more - unhandled exception error is happening there (put a separate issue in for that).

Also, when you do choose an image, the file name doesn't show except for the first letter. Could this also be changed? This is when you are still on the Create activity page.

![image](https://cloud.githubusercontent.com/assets/17350634/13976710/15162ed2-f088-11e5-897e-552ca8cf4cf2.png)

Hope this helps!

**Comment by dangle1** [comment]
Incidentally, I'm seeing even more unexpected behavior. It appears adding a file causes the server to bypass all other validations on the page besides the date validations. So far I've been able to submit the form without a name for the activity, activity type, or volunteers requested; because there's a constraint in the db for activity name, I get an exception page, but entering a name allows me to create the activity without the other two required fields. 

Here's a gif of the steps to reproduce. 

![activity image bug](https://cloud.githubusercontent.com/assets/3221548/13977027/b3ebd6ca-f082-11e5-8560-1c7f061e9362.gif)

Unfortunately I don't have "Organizer" rights on the live site so I can't actually test this anywhere else other than my machine. @palittle can you confirm please? I'll spin up another issue for these findings and close it if it's not happening on the live site.

**Comment by dangle1** [comment]
@palittle 

Just saw your latest reply, and acknowledge. I think the label for the filename is dependent on the window size/zoom, so try maximizing your window and ensuring your zoom is at %100.

**Comment by dangle1** [comment]
So it turns out that this behavior is somewhat expected: 

Browsers can't prepopulate input tags for files ([this is for security reasons](http://stackoverflow.com/questions/967916/how-to-keep-input-type-file-field-value-after-failed-validation-in-asp-net-mvc)). So when the form is submitted and fails validation, the server redirects to the same page, thus resulting in the condition above. 

There are individuals out there who have managed to circumvent this constraint, but the implementations venture into 'hacky' territory, and some point, you have to wonder if the juice is worth the squeeze. 

That is not to say that we have to rely on the server-side validations, however. It is possible to wire up the same validations on the client side so that the form doesn't have to reach the server to get validated in the first place. This would keep the browser on the same page, display the necessary validation messages, and keep the file name in place so that the user doesn't have to reselect the image. 

I'd be happy to implement that solution but would like some input before doing so as I don't want to set an unintended precedence (or expend unnecessary energy in an attempt to do so). MVC comes with `jquery.validate.unobtrusive` which is extensible to custom client side validations. However, the project is also using `knockout.js` which has `knockout-validation.js`. Both approaches are comparable for implementing the above, but once again, would like to know the preference from the community before going in one direction or the other. 

@tonysurma, do you have any opinions on this or know of who should be aware of this decision?

**Comment by tonysurma** [comment]
pinging @MisterJames @mheggeseth @dpaquette @BillWagner for add'l thoughts but here are mine:

I would say that we should have client side validation for everything that can be client side validated (thought this would be there already)

For when/if it needs to go to the server then yeah we lose the image per the comments and agree that we can't pursue a fix.

As for jquery.validate vs. knockout I would assume we are suing one for client side already (bad assumption?) and therefore point you to that as the choice as well as any comments from the folks pinged above.  Thanks
